text/0102-offchain-parachain-runtime-upgrades.md

```diff
@@ -0,0 +1,354 @@
+   blocks.
+
+We discussed introducing a separate signalling before submitting the actual
+runtime, but I think we should just go one step further and make upgrades fully
+off-chain. Which also helps bringing down deposit costs in a secure way, as we
+are also actually reducing costs for the network.
+
+### Introduce a new UMP message type `RequestCodeUpgrade`
+
+As part of elastic scaling we are already planning to increase flexibility of [UMP
+messages](https://github.com/polkadot-fellows/RFCs/issues/92#issuecomment-2144538974), we can now use this to our advantage and introduce another UMP message:
+
+```rust
+enum UMPSignal {
+  // For elastic scaling
+  OnCore(CoreIndex),
+  // For off-chain upgrades
+  RequestCodeUpgrade(Hash),
+}
+```
+
+We could also make that new message a regular XCM, calling an extrinsic on the
+relay chain, but we will want to look into that message right after validation
+on the backers on the node side, making a straight forward semantic message more
+apt for the purpose.
+
+
+### Handle `RequestCodeUpgrade` on backers
+
+We will introduce a new request/response protocol for both collators and
+validators, with the following request/response:
+
+```rust
+struct RequestBlob {
+  blob_hash: Hash,
+}
+
+struct BlobResponse {
+  blob: Vec<u8>
+}
+```
+
+This protocol will be used by backers to request the PVF from collators in the
+following conditions:
+
+1. They received a collation sending `RequestCodeUpgrade`.
+2. They received a collation, but they don't yet have the code that was
+   previously registered on the relaychain. (E.g. disk pruned, new validator)
+
+In case they received the collation via PoV distribution instead of from the
+collator itself, they will use the exact same message to fetch from the valiator
+they got the PoV from.
+
+### Get the new code to all validators
+
+Once the candidate issuing `RequestCodeUpgrade` got backed on chain, validators
+will start fetching the code from the backers as part of availability
+distribution.
+
+To mitigate attack vectors we should make sure that serving requests for code
+can be treated as low priority requests. Thus I am suggesting the following
+scheme:
+
+Validators will notice via a runtime API (TODO: Define) that a new code has been requested, the
+API will return the `Hash` and a counter, which starts at some configurable
+value e.g. 10. The validators are now aware of the new hash and start fetching,
+but they don't have to wait for the fetch to succeed to sign their bitfield.
+
+Then on each further candidate from that chain that counter gets decremented.
+Validators which have not yet succeeded fetching will now try again. This game
+continues until the counter reached `0`. Now it is mandatory to have to code in
+order to sign a `1` in the bitfield.
+
+PVF pre-checking will happen after the candidate which brought the counter to
+`0` has been successfully included and thus is also able to assume that 2/3 of
```

Comment:
Is there an expiry date for when the parachain needs to reach 0, otherwise the code upgrade is dropped ?

Reply:
Good point. Will add a a section.